File: Problems/Algorithms/3. Longest Substring without Repeating Characters/longest_substring_no_repeat.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        if n <= 1:
            return n

        count = {}

        i, j, ans = 0, 0, 0
        count[s[0]] = 1
        while j < n-1:
            if s[j+1] not in count or count[s[j+1]] == 0:
                count[s[j+1]] = 1
                j += 1
            else:
                count[s[i]] -= 1
                i += 1
            ans = max(ans, j-i+1)

        return ans

File: Problems/Algorithms/3. Longest Substring without Repeating Characters/longest_substring_no_repeat.rs
use std::collections::HashMap;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut start: usize = 0;
        let mut ans = 0;
        let mut map: HashMap<char, usize> = HashMap::new();

        for (idx, ch) in s.char_indices() {
            map.entry(ch).and_modify(|prev| {
                if *prev >= start {
                    ans = ans.max(idx-start);
                    start = *prev + 1;
                }
                *prev = idx;
            }).or_insert(idx);
        }

        ans.max(s.len()-start) as i32
    }
}

